### PR TITLE
FadeInImage cacheWidth and cacheHeight support

### PR DESCRIPTION
**Pull request by GaryQian**
## Description

This adds support for `cacheWidth` and `cacheHeight` in `FadeInImage`. This is a continuation of https://github.com/flutter/flutter/pull/41415 and https://github.com/flutter/flutter/issues/26194

## Related Issues

https://github.com/flutter/flutter/issues/26194

This change was requested in https://github.com/flutter/flutter/pull/42785.

### CODE REVIEW

**Comment by GaryQian** [comment]
Tests in progress.

File: packages/flutter/lib/src/widgets/fade_in_image.dart [hunk 1]
```diff
@@ -16,6 +16,14 @@ import 'transitions.dart';
 // Examples can assume:
 // Uint8List bytes;
 
+// Same as in painting/image.dart.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/fade_in_image.dart]: Can we put this in a common place and reuse it?

File: packages/flutter/lib/src/widgets/fade_in_image.dart [hunk 2]
```diff
@@ -137,6 +158,10 @@ class FadeInImage extends StatelessWidget {
     this.alignment = Alignment.center,
     this.repeat = ImageRepeat.noRepeat,
     this.matchTextDirection = false,
+    int placeholderCacheWidth,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/fade_in_image.dart]: Why not use two Size arguments instead of four ints?

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/fade_in_image.dart]: Ah, I see. The other Images also expose it as separate ints. I guess we should be consistent.

File: packages/flutter/lib/src/widgets/fade_in_image.dart [hunk 3]
```diff
@@ -66,6 +74,12 @@ class FadeInImage extends StatelessWidget {
   /// Creates a widget that displays a [placeholder] while an [image] is loading,
   /// then fades-out the placeholder and fades-in the image.
   ///
+  /// To include a custom decode/cache size such as those provided by
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/fade_in_image.dart]: nit: This sentence is very complicated and hard to parse. What are you trying to say?

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/fade_in_image.dart]: Isn't this basically just saying that placeholder and image may also be ResizeImages?

File: packages/flutter/lib/src/widgets/fade_in_image.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/fade_in_image.dart]: Maybe it could be a static method on ImageProvider?

File: packages/flutter/test/widgets/fade_in_image_test.dart [hunk 4]
```diff
@@ -262,6 +265,44 @@ Future<void> main() async {
...
+      );
+
+      final DecoderCallback decode = (Uint8List bytes, {int cacheWidth, int cacheHeight}) {
+        expect(cacheWidth, null);
+        expect(cacheHeight, null);
+        return PaintingBinding.instance.instantiateImageCodec(bytes, cacheWidth: cacheWidth, cacheHeight: cacheHeight);
+      };
+      // image.placeholder should be an instance of MemoryImage instead of ResizeImage
+      final MemoryImage memoryImage = image.placeholder;
+      expect(image.placeholder is MemoryImage, true);
```
Comment (goderbauer) [inline packages/flutter/test/widgets/fade_in_image_test.dart]: `expect(image.placeholder, isA<MemoryImage>)`

Also, the type in the line above should probably be ImageProvider instead of MemoryImage. Otherwise you may never get to this expect.

File: packages/flutter/test/widgets/fade_in_image_test.dart [hunk 5]
```diff
@@ -262,6 +265,44 @@ Future<void> main() async {
...
+        imageCacheWidth: 40,
+        imageCacheHeight: 50,
+      );
+
+      final DecoderCallback decode = (Uint8List bytes, {int cacheWidth, int cacheHeight}) {
+        expect(cacheWidth, 20);
+        expect(cacheHeight, 30);
+        return PaintingBinding.instance.instantiateImageCodec(bytes, cacheWidth: cacheWidth, cacheHeight: cacheHeight);
+      };
+      final ResizeImage resizeImage = image.placeholder;
```
Comment (goderbauer) [inline packages/flutter/test/widgets/fade_in_image_test.dart]: Should we have an expect about the type here as well like in the other test?

File: packages/flutter/test/widgets/fade_in_image_test.dart [hunk 6]
```diff
@@ -262,6 +265,44 @@ Future<void> main() async {
...
+      resizeImage.load(await resizeImage.obtainKey(ImageConfiguration.empty), decode);
```
Comment (goderbauer) [inline packages/flutter/test/widgets/fade_in_image_test.dart]: It makes me a bit uncomfortable that the test doesn't verify that the decode callback was actually invoked. If for some reason it doesn't run, the expects in there also wouldn't execute.

File: packages/flutter/lib/src/widgets/fade_in_image.dart (hunk 1 above)
Comment (GaryQian) [reply packages/flutter/lib/src/widgets/fade_in_image.dart]: Moved to a static on ResizeImage

File: packages/flutter/lib/src/widgets/fade_in_image.dart (hunk 3 above)
Comment (GaryQian) [reply packages/flutter/lib/src/widgets/fade_in_image.dart]: Added a simplified explanation

File: packages/flutter/test/widgets/fade_in_image_test.dart (hunk 6 above)
Comment (GaryQian) [reply packages/flutter/test/widgets/fade_in_image_test.dart]: Verifies it is called now!

File: packages/flutter/test/widgets/fade_in_image_test.dart [hunk 7]
```diff
@@ -84,6 +84,10 @@ FadeInImageParts findFadeInImage(WidgetTester tester) {
   }
 }
 
+class ClosureTracker {
```
Comment (goderbauer) [inline packages/flutter/test/widgets/fade_in_image_test.dart]: You don't really need an extra class for this. Just define a bool variable outside the closure, initialize it with false and set it to true within the closure.

File: packages/flutter/test/widgets/fade_in_image_test.dart [hunk 8]
```diff
@@ -279,10 +283,15 @@ Future<void> main() async {
       final DecoderCallback decode = (Uint8List bytes, {int cacheWidth, int cacheHeight}) {
         expect(cacheWidth, 20);
         expect(cacheHeight, 30);
+        ClosureTracker.called = true;
         return PaintingBinding.instance.instantiateImageCodec(bytes, cacheWidth: cacheWidth, cacheHeight: cacheHeight);
       };
-      final ResizeImage resizeImage = image.placeholder;
-      resizeImage.load(await resizeImage.obtainKey(ImageConfiguration.empty), decode);
+      final ImageProvider resizeImage = image.placeholder;
+      expect(image.placeholder is ResizeImage, true);
```
Comment (goderbauer) [inline packages/flutter/test/widgets/fade_in_image_test.dart]: expect(image.placeholder, isA<ResizeImage>());

File: packages/flutter/test/widgets/fade_in_image_test.dart [hunk 9]
```diff
@@ -295,12 +304,16 @@ Future<void> main() async {
       final DecoderCallback decode = (Uint8List bytes, {int cacheWidth, int cacheHeight}) {
         expect(cacheWidth, null);
         expect(cacheHeight, null);
+        ClosureTracker.called = true;
         return PaintingBinding.instance.instantiateImageCodec(bytes, cacheWidth: cacheWidth, cacheHeight: cacheHeight);
       };
       // image.placeholder should be an instance of MemoryImage instead of ResizeImage
-      final MemoryImage memoryImage = image.placeholder;
+      final ImageProvider memoryImage = image.placeholder;
       expect(image.placeholder is MemoryImage, true);
```
Comment (goderbauer) [inline packages/flutter/test/widgets/fade_in_image_test.dart]: expect(image.placeholder, isA< MemoryImage>());

File: packages/flutter/test/widgets/fade_in_image_test.dart (hunk 7 above)
Comment (GaryQian) [reply packages/flutter/test/widgets/fade_in_image_test.dart]: Hmm I tried this initially and it didnt seem to work, I probably did it wrong. I'll try again.

Comment (GaryQian) [reply packages/flutter/test/widgets/fade_in_image_test.dart]: Ok it works now, idk what I did earlier for this to not work..

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/fade_in_image.dart (hunk 1 above)
Comment (nightfallsad) [reply packages/flutter/lib/src/widgets/fade_in_image.dart]: > Maybe it could be a static method on ImageProvider?

Will this PR merge to the current last stable version v1.9.1+hotfix.6？
